src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberMethodSymbol.cs

```diff
@@ -219,6 +220,16 @@ private void MethodChecks(MethodDeclarationSyntax syntax, Binder withTypeParamsB
                     var loc = parameterSyntax.Type.Location;
                     diagnostics.Add(ErrorCode.ERR_BadTypeforThis, loc, parameter0Type);
                 }
+                else if (parameter0RefKind == RefKind.Ref && !parameter0Type.IsValueType)
+                {
+                    diagnostics.Add(ErrorCode.ERR_RefExtensionMustBeValueTypeOrConstrainedToOne, location, Name);
+                    this.SetIsExtensionMethod(false);
```

Comment:
Also, it looks like we are mutating the state of the method that is supposed to stay immutable. It looks very possible that one can access the IsExtensionMethod property before you mutate it here.